api/authenticator.py

```diff
@@ -232,69 +579,704 @@ def create_authentication_headers(self):
+
+        This method should call create_token() at some point.
+
+        :return: A ProblemDetail if there's a problem. Otherwise, a
+        3-tuple (Credential, Patron, PatronData). The Credential
+        contains the access token provided by the OAuth
+        provider. (This is not the bearer token the patron will use to
+        authenticate -- that is a JWT token _based_ on this token,
+        created in
+        OAuthAuthenticationController.oauth_authentication_callback). The
+        Patron object represents the authenticated Patron, and the
+        PatronData object may include information about the patron
+        obtained from the OAuth provider which cannot be stored in the
+        circulation manager's database but which should be passed on
+        to the client.
 
-    def patron_info(self, identifier):
-        return {}
+        """
+        raise NotImplementedError()
 
     def _internal_authenticate_url(self):
-        return url_for('oauth_authenticate', _external=True, provider=self.NAME)
+        """A patron who wants to log in should hit this URL on the circulation
+        manager. They'll be redirected to the OAuth provider, which will 
+        take care of it.
+        """
+        return url_for('oauth_authenticate', _external=True,
+                       provider=self.NAME)
 
-    def create_authentication_provider_document(self):
+    @property
+    def authentication_provider_document(self):
+        """Create a stanza for use in an Authentication for OPDS document.
+
+        Example:
+        {
+            "name": "My OAuth Provider",
+            "methods": {
+                "http://librarysimplified.org/authtype/MyOAuthProvider" : {
+                "links": {
+                    "authenticate": "https://circulation.library.org/oauth_authenticate?provider=MyOAuth"
+                 }
+            }
+        }
+        """
         method_doc = dict(links=dict(authenticate=self._internal_authenticate_url()))
         methods = {}
         methods[self.METHOD] = method_doc
         return dict(name=self.NAME, methods=methods)
+
+    def token_data_source(self, _db):
+        # TODO: This is one of those situations where we'd like to
+        # create the data source if it doesn't already exist.
+        return DataSource.lookup(_db, self.TOKEN_DATA_SOURCE_NAME)
+
+
+class OAuthController(object):
+
+    """A controller for handling requests that are part of the OAuth
+    credential dance.
+    """
+    
+    def __init__(self, authenticator):
+        self.authenticator = authenticator
+
+    def oauth_authentication_redirect(self, params):
+        """Redirect an unauthenticated patron to the authentication URL of the
+        appropriate OAuth provider.
+
+        Over on that other site, the patron will authenticate and be
+        redirected back to the circulation manager (the URL is stored
+        in params['redirect_uri']), ending up in
+        oauth_authentication_callback.
```

Comment:
I've changed the comment, and refactored the external_authenticate_url code in a way that I hope addresses this issue.